asf_heat_pump_suitability/pipeline/prepare_features/listed_buildings.py

```diff
@@ -29,56 +29,95 @@ def transform_df_EPC_X_and_Y_to_point(
+def get_filtered_df_listed_buildings(data_source_key: str) -> gpd.GeoDataFrame:
     """
     Get and filter out listed buildings from Historic England dataset.
 
     Returns:
         gpd.GeoDataFrame: Filtered Historic England dataset with only listed buildings grade and geometry.
     """
-    historic_england_gdf = get_df_historicengland_listedbuildings()
+    listed_building_gdf = get_df_listedbuildings(data_source_key)
     relevant_columns = ["geometry", "Grade"]
-    filtered_historic_england_gdf = historic_england_gdf[relevant_columns]
+    filtered_listed_building_gdf = listed_building_gdf[relevant_columns]
 
-    return filtered_historic_england_gdf
+    return filtered_listed_building_gdf
 
 
 def spatial_join_epc_with_listed_buildings(
-    enhanced_epc_df: pl.DataFrame, listed_buildings_df: gpd.GeoDataFrame
-) -> gpd.GeoDataFrame:
+    enhanced_epc_df: pl.DataFrame,
+    listed_buildings_df: gpd.GeoDataFrame,
+    chunk_size: int = 100000,
+) -> pl.DataFrame:
     """
     Spatial join EPC dataset with listed buildings dataset.
     Args:
         enhanced_epc_df (pl.DataFrame): Enhanced EPC dataset
         listed_buildings_df (gpd.GeoDataFrame): Filtered Historic England dataset with only listed buildings grade and geometry.
     Returns:
-        gpd.GeoDataFrame: EPC dataset with listed buildings grade without geometry.
+        pl.DataFrame: EPC dataset with listed buildings grade without geometry.
     """
-    epc_gdf = transform_df_EPC_X_and_Y_to_point(enhanced_epc_df)
-    epc_gdf_temp = epc_gdf[["geometry"]].copy()
-    epc_gdf_temp["index"] = epc_gdf.index
-    joined_gdf = gpd.sjoin(
-        epc_gdf_temp, listed_buildings_df, how="left", predicate="intersects"
+
+    # Add a unique index column for merging later
+    enhanced_epc_df = enhanced_epc_df.with_row_index("index")
+    partitions = enhanced_epc_df.with_row_count("chunk_id").select(
+        pl.col("chunk_id") // chunk_size
     )
-    # Drop the geometry column
-    joined_gdf = joined_gdf.drop(columns=["geometry", "index_right"])
-    epc_gdf["index"] = epc_gdf.index
-    result_gdf = epc_gdf.merge(
-        joined_gdf[["Grade", "index"]], on="index", how="left"
-    ).drop(columns=["geometry", "index"])
-    return result_gdf
+    # group based on the created index, resulting chunk_size partitons
+    data_partitioned = enhanced_epc_df.with_columns(partitions).partition_by("chunk_id")
+    logging.info(f"Adding listed buildings to EPC in {len(data_partitioned)} chunks")
+    for i, enhanced_epc_df_chunk in tqdm(enumerate(data_partitioned)):
+        epc_gdf = transform_df_EPC_X_and_Y_to_point(enhanced_epc_df_chunk)
+        epc_gdf_temp = epc_gdf[["geometry", "index"]].copy()
+        joined_gdf_chunk = gpd.sjoin(
+            epc_gdf_temp, listed_buildings_df, how="inner", predicate="intersects"
+        )
+        # Drop the geometry column
+        joined_gdf_chunk = joined_gdf_chunk.drop(columns=["geometry", "index_right"])
```

Comment:
```suggestion
        joined_gdf_chunk = gpd.sjoin(
            epc_gdf_temp, listed_buildings_df, how="inner", predicate="intersects"
        ).drop(columns=["geometry", "index_right"])
```
Minor but I think you could chain this if you wanted to.

Reply:
Cool, happy with this!